core/lib/src/rocket.rs

```diff
@@ -641,135 +785,76 @@ impl Rocket {
-            let rocket = rocket.clone();
-            let remote_addr = connection.remote_addr().unwrap_or_else(|| ([0, 0, 0, 0], 0).into());
-            async move {
-                Ok::<_, std::convert::Infallible>(hyper::service_fn(move |req| {
-                    hyper_service_fn(rocket.clone(), remote_addr, req)
-                }))
-            }
-        });
-
-        #[derive(Clone)]
-        struct TokioExecutor;
-
-        impl<Fut> hyper::Executor<Fut> for TokioExecutor where Fut: Future + Send + 'static, Fut::Output: Send {
-            fn execute(&self, fut: Fut) {
-                tokio::spawn(fut);
+    pub(crate) fn finish(&mut self) -> BoxFuture<'_, ()> {
+        Box::pin(async move {
+            while !self.pending.is_empty() {
+                let op = self.pending.remove(0);
+                let inner = self.take_inner();
+                self.inner = Some(match op {
+                    BuildOperation::Mount(base, routes) => inner._mount(base, routes),
+                    BuildOperation::Register(catchers) => inner._register(catchers),
+                    BuildOperation::Manage(callback) => inner._manage(callback),
+                    BuildOperation::Attach(fairing) => inner._attach(fairing).await,
+                });
             }
-        }
-
-        hyper::Server::builder(Incoming::from_listener(listener))
-            .executor(TokioExecutor)
-            .serve(service)
-            .with_graceful_shutdown(async move { shutdown_receiver.recv().await; })
-            .await
-            .map_err(|e| crate::error::Error::Run(Box::new(e)))
+        })
     }
 
-    /// Returns a `Future` that drives the server and completes when the server
-    /// is shut down or errors. If the `ctrl_c_shutdown` feature is enabled,
-    /// the server will shut down gracefully once `Ctrl-C` is pressed.
+    /// Returns a `Future` that drives the server, listening for and dispathcing
+    /// requests to mounted routes and catchers. The `Future` completes when the
+    /// server is shut down (via a [`ShutdownHandle`] or encounters a fatal
+    /// error.  If the `ctrl_c_shutdown` feature is enabled, the server will
+    /// also shut down once `Ctrl-C` is pressed.
+    ///
+    /// # Error
+    /// If there is a problem starting the application, an [`Error`] is
+    /// returned. Note that a value of type `Error` panics if dropped
+    /// without first being inspected. See the [`Error`] documentation for
+    /// more information.
     ///
     /// # Example
     ///
     /// ```rust
-    /// #[tokio::main]
+    /// #[rocket::main]
     /// async fn main() {
     /// # if false {
-    ///     let result = rocket::ignite().serve().await;
+    ///     let result = rocket::ignite().launch().await;
     ///     assert!(result.is_ok());
     /// # }
     /// }
     /// ```
-    pub async fn serve(self) -> Result<(), crate::error::Error> {
+    pub async fn launch(mut self) -> Result<(), crate::error::Error> {
```

Comment:
I know most the contents of this function were introduced in some other commits, but it has become very difficult to follow. The combination of macros and `cfg`s, in particular, really help to obfuscate the functionality. We really must improve this before we release 0.5.